@jurakp  Please fix these errors : 
```
cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/PropertyOrderException.java:27: assign child at indentation level 4 not at correct indentation, 8
/home/jimma/code/cxf-cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/PropertyOrderException.java:28: assign child at indentation level 0 not at correct indentation, 8
Audit done.
```
and
```
ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[31,34] package org.apache.cxf.ext.logging does not exist
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[32,40] package org.apache.cxf.ext.logging.event does not exist
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[33,40] package org.apache.cxf.ext.logging.event does not exist
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[96,52] cannot find symbol
  symbol:   class LogEventSender
  location: class org.apache.cxf.systest.jaxb.TestServiceTest
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[106,26] cannot find symbol
  symbol:   class LogEvent
  location: class org.apache.cxf.systest.jaxb.TestServiceTest.TestLoggingEventSender
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[82,84] cannot find symbol
  symbol:   class LoggingOutInterceptor
  location: class org.apache.cxf.systest.jaxb.TestServiceTest
[ERROR] cxf/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java:[105,9] method does not override or implement a method from a supertype
```

--- a/rt/databinding/jaxb/src/main/java/org/apache/cxf/jaxb/JAXBEncoderDecoder.java
+++ b/rt/databinding/jaxb/src/main/java/org/apache/cxf/jaxb/JAXBEncoderDecoder.java
@@ -56,6 +56,7 @@ import javax.xml.bind.annotation.XmlAccessOrder;
 import javax.xml.bind.annotation.XmlAccessType;
 import javax.xml.bind.annotation.XmlAccessorOrder;
 import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlType;
 import javax.xml.bind.annotation.adapters.HexBinaryAdapter;
 import javax.xml.bind.attachment.AttachmentMarshaller;
 import javax.xml.bind.attachment.AttachmentUnmarshaller;
@@ -425,7 +426,28 @@ public final class JAXBEncoderDecoder {
                     }
                 });
             }
-
+            XmlType xmlType = cls.getAnnotation(XmlType.class);
+            if (xmlType != null && xmlType.propOrder().length > 1 && !xmlType.propOrder()[0].isEmpty()) {
+                final List<String> orderList = Arrays.asList(xmlType.propOrder());
+                Collections.sort(combinedMembers, new Comparator<Member>() {
+                    public int compare(Member m1, Member m2) {
+                        String m1Name = getName(m1);
+                        String m2Name = getName(m2);
+                        int m1Index = orderList.indexOf(m1Name);
+                        int m2Index = orderList.indexOf(m2Name);
+                        if (m1Index != -1 && m2Index != -1) {
+                            return m1Index - m2Index;
+                        }
+                        if (m1Index == -1 && m2Index != -1) {
+                            return 1;
+                        }
+                        if (m1Index != -1 && m2Index == -1) {
+                            return -1;
+                        }
+                        return 0;
+                    }
+                });
+            }
             for (Member member : combinedMembers) {
                 if (member instanceof Field) {
                     Field f = (Field)member;
@@ -460,8 +482,20 @@ public final class JAXBEncoderDecoder {
             StaxUtils.close(writer);
         }
     }
-    
-    private static void writeArrayObject(Marshaller marshaller, 
+
+    private static String getName(Member m1) {
+        String m1Name = null;
+        if (m1 instanceof Field) {
+            m1Name = ((Field)m1).getName();
+        } else {
+            int idx = m1.getName().startsWith("get") ? 3 : 2;
+            String name = m1.getName().substring(idx);
+            m1Name = Character.toLowerCase(name.charAt(0)) + name.substring(1);
+        }
+        return m1Name;
+    }
+
+    private static void writeArrayObject(Marshaller marshaller,
                                          Object source,
                                          QName mname,
                                          Object mObj) throws Fault, JAXBException {

--- a/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java
+++ b/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/TestServiceTest.java
@@ -19,6 +19,8 @@
 package org.apache.cxf.systest.jaxb;
 
 
+import java.io.ByteArrayOutputStream;
+import java.io.PrintWriter;
 import java.net.URL;
 
 import javax.xml.namespace.QName;
@@ -29,8 +31,10 @@ import org.w3c.dom.Document;
 import org.apache.cxf.Bus;
 import org.apache.cxf.endpoint.Server;
 import org.apache.cxf.helpers.IOUtils;
+import org.apache.cxf.interceptor.LoggingOutInterceptor;
 import org.apache.cxf.systest.jaxb.model.ExtendedWidget;
 import org.apache.cxf.systest.jaxb.model.Widget;
+import org.apache.cxf.systest.jaxb.service.PropertyOrderException;
 import org.apache.cxf.systest.jaxb.service.TestService;
 import org.apache.cxf.test.TestUtilities;
 import org.apache.cxf.testutil.common.TestUtil;
@@ -44,7 +48,6 @@ import org.springframework.test.context.junit4.AbstractJUnit4SpringContextTests;
 @ContextConfiguration(locations = { "classpath:extrajaxbclass.xml" })
 public class TestServiceTest extends AbstractJUnit4SpringContextTests {
     static final String PORT = TestUtil.getPortNumber(TestServiceTest.class);
-
     private TestUtilities testUtilities;
 
     public TestServiceTest() {
@@ -73,7 +76,28 @@ public class TestServiceTest extends AbstractJUnit4SpringContextTests {
         Assert.assertEquals(expected, widgetFromService);
     }
 
-    
+    @Test
+    public void testExceptionPropertyOrder() throws Throwable {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        PrintWriter pw = new PrintWriter(baos, true);
+        ((Bus)applicationContext.getBean("cxf")).getOutFaultInterceptors().add(new LoggingOutInterceptor(pw));
+        TestService testClient = getTestClient();
+        ((BindingProvider)testClient).getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY,
+                                                              "http://localhost:" + PORT
+                                                                  + "/service/TestEndpoint");
+        try {
+            testClient.echo("Exception");
+        } catch (PropertyOrderException e) {
+            String payload = baos.toString();
+            Assert.assertTrue("Expect <message> element is before <data> element :" + payload,
+                              payload.indexOf("</message><data") > -1);
+        } finally {
+            if (pw != null) {
+                pw.close();
+            }
+        }
+    }
+
     @Test
     public void testSchema() throws Exception {
         URL url = new URL("http://localhost:" + PORT + "/service/TestService?wsdl");

--- a/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/ErrorData.java
+++ b/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/ErrorData.java
@@ -1,0 +1,62 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.cxf.systest.jaxb.service;
+
+import javax.xml.bind.annotation.XmlType;
+
+@XmlType(name = "ErrorData")
+public class ErrorData implements java.io.Serializable {
+    private static final long serialVersionUID = 1L;
+    private long code;
+    private String title;
+    private String description;
+
+    public ErrorData() {
+    }
+
+    public ErrorData(long code, String title, String description, String stack) {
+        this.code = code;
+        this.title = title;
+        this.description = description;
+    }
+
+    public long getCode() {
+        return code;
+    }
+
+    public void setCode(long code) {
+        this.code = code;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public void setDescription(String description) {
+        this.description = description;
+    }
+}

--- a/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/PropertyOrderException.java
+++ b/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/PropertyOrderException.java
@@ -1,0 +1,70 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.cxf.systest.jaxb.service;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlType;
+
+@XmlRootElement(name = "fault")
+@XmlType(name = "", propOrder = {
+        "message", "data"
+        }, namespace = "")
+public class PropertyOrderException extends Exception implements java.io.Serializable {
+
+    private static final long serialVersionUID = 1L;
+
+    @XmlElement(name = "message", required = true, nillable = true)
+    private String message;
+
+    @XmlElement(name = "data", required = true, nillable = true)
+    private ErrorData data;
+
+    public PropertyOrderException() {
+        this.data = new ErrorData();
+        this.message = null;
+    }
+
+    public PropertyOrderException(ErrorData data) {
+        this.data = data;
+        this.message = null;
+    }
+
+    public PropertyOrderException(String message) {
+        this.data = null;
+        this.message = message;
+    }
+
+    public String getMessage() {
+        return this.message;
+    }
+
+    public ErrorData getData() {
+        return data;
+    }
+
+    public void setData(ErrorData data) {
+        this.data = data;
+    }
+
+    public void setMessage(String message) {
+        this.message = message;
+    }
+
+}

--- a/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/TestService.java
+++ b/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/TestService.java
@@ -34,5 +34,8 @@ public interface TestService {
     
     @WebMethod
     void testExceptionMarshalling() throws TestServiceException;
+    
+    @WebMethod
+    String echo(String test) throws PropertyOrderException;
 
 }

--- a/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/TestServiceImpl.java
+++ b/systests/databinding/src/test/java/org/apache/cxf/systest/jaxb/service/TestServiceImpl.java
@@ -19,6 +19,7 @@
 
 package org.apache.cxf.systest.jaxb.service;
 
+import javax.jws.WebMethod;
 import javax.jws.WebService;
 
 import org.apache.cxf.systest.jaxb.model.ExtendedWidget;
@@ -35,6 +36,20 @@ public class TestServiceImpl implements TestService {
         throw new TestServiceException("Your hovercraft is full of eels.");
     }
 
-    
+    @WebMethod
+    public String echo(String test) throws PropertyOrderException {
+        if ("Exception".equalsIgnoreCase(test)) {
+            PropertyOrderException exception = new PropertyOrderException();
+            ErrorData ed = new ErrorData();
+            ed.setCode(500);
+            ed.setDescription("Error happened");
+            ed.setTitle("Error title");
+            exception.setData(ed);
+            exception.setMessage("Exception message");
+            throw exception;
+
+        }
+        return test;
+    }
 
 }

